### PR TITLE
39 permissions for task pallet

### LINKED ISSUE #39: Permissions for Task Pallet
Things that we have to be improved:

- [x]  The creator of a task should not be the volunteer.
- [x]  Task can be deleted by the creator only after it has been created.
- [x]  Ensure all proper TaskStatus checks before making a transition.
- [x]  Ensure all Task ownership checks before transferring task ownership.
- [x]  Perpahs store tasks with passed deadlines to make sure they can be edited or deleted.
- [x]  Re-examine how we deal with permissions.

Budget: $70

### PR DESCRIPTION
**Pull request by stojanov-igor**
Enforced stronger permissions
Added remove_task extrinsic
Updated Benchmark & weights
Added tests for new functionality

Budget: $15

### CODE REVIEW

File: pallets/task/src/lib.rs [hunk 1]
```diff
@@ -42,22 +42,49 @@
 //!
 //! Furthermore, budget funds are locked in escrow when task is created.
 //! Funds are removed from escrow when task is removed.
+//! 
+//! Tasks with expired deadline are automatically removed from storage.
```
Comment (vivekvpandya) [inline pallets/task/src/lib.rs:46]: Just to Note: we must track such task as once removed there is no way to get that back. So subsquid seems better solution as it provides GraphQL interface for query. Other wise some private instance of DB.

File: pallets/task/src/lib.rs [hunk 2]
```diff
@@ -252,16 +288,34 @@ pub mod pallet {
...
 		}
 
+		/// Function that removes a task by task owner. [origin, task_id]
+		#[pallet::weight(<T as Config>::WeightInfo::remove_task(0,0))]
+		pub fn remove_task(origin: OriginFor<T>, task_id: T::Hash) -> DispatchResult {
+
+			// Check that the extrinsic was signed and get the signer.
+			let signer = ensure_signed(origin)?;
+
+			// Assign task and update storage.
```
Comment (vivekvpandya) [inline pallets/task/src/lib.rs]: Comment needs update.

File: pallets/task/src/lib.rs [hunk 3]
```diff
@@ -252,16 +288,34 @@ pub mod pallet {
...
+			Self::delete_task(&signer, &task_id)?;
+
+			// Emit a Task Assigned Event.
```
Comment (vivekvpandya) [inline pallets/task/src/lib.rs]: Comment needs update.

File: pallets/task/src/lib.rs [hunk 4]
```diff
@@ -499,7 +563,8 @@ pub mod pallet {
 			Ok(())
 		}
 
-		pub fn delete_task(task_initiator: &T::AccountId, task_id: &T::Hash) -> Result<(), DispatchError> {
+		pub fn accept_completed_task(task_initiator: &T::AccountId, task_id: &T::Hash) -> Result<(), DispatchError> {
+			
 			// Check if task exists
 			let mut task = Self::tasks(&task_id).ok_or(<Error<T>>::TaskNotExist)?;
```
Comment (vivekvpandya) [inline pallets/task/src/lib.rs:570]: is_task_initiator() checks for task exists or not. So perphas Self::tasks() can be removed and just directly mutate task after is_task_initiator() check.

**Review by vivekvpandya — APPROVED** [review]
Rest of the things looks good.

File: pallets/task/src/lib.rs (hunk 1 above)
Comment (stojanov-igor) [reply pallets/task/src/lib.rs:46]: Created a separate task for this. https://github.com/UniversalDot/universal-dot-node/issues/46 

It is better not to use DB, but rather store them in a separate vector array for (2 weeks) after which you will permanently delete them.

File: pallets/task/src/lib.rs (hunk 4 above)
Comment (stojanov-igor) [reply pallets/task/src/lib.rs:570]: Good catch... However, we are using `let mut task`, to get the current task struct and do mutations on it.

**Pull request merged**